go/pools/fast_pool.go

```diff
@@ -0,0 +1,604 @@
+			if !ok {
+				p.Lock()
+				p.state.Waiters--
+				p.Unlock()
+
+				return nil, ErrClosed
+			}
+
+			p.Lock()
+			p.state.InPool--
+			p.state.InUse++
+			p.state.Waiters--
+			p.state.WaitCount++
+			p.state.WaitTime += time.Now().Sub(startTime)
+			p.Unlock()
+
+			return wrapper.resource, nil
+
+		case <-ctx.Done():
+			p.Lock()
+			p.state.Waiters--
+			p.Unlock()
+
+			return nil, ErrTimeout
+
+		case <-time.After(100 * time.Millisecond):
+			// There could be a condition where this caller has been
+			// put into a queue, but another caller has failed in creating
+			// a resource, causing a deadlock. We'll check occasionally to see
+			// if there is now capacity to create.
+			if p.State().Closed {
+				return nil, ErrClosed
+			}
+			wrapper, err := p.safeCreate()
+			if err == errNeedToQueue {
+				continue
+			} else if err != nil {
+				return nil, err
+			} else {
+				return wrapper.resource, nil
+			}
+		}
+	}
+}
+
+// Put will return a resource to the pool. For every successful Get,
+// a corresponding Put is required. If you no longer need a resource,
+// you will need to call Put(nil) instead of returning the closed resource.
+// The will eventually cause a new resource to be created in its place.
+func (p *FastPool) Put(resource Resource) {
+	p.Lock()
+
+	p.state.InUse--
+
+	if p.state.Closed || p.active() > p.state.Capacity {
+		// We're ether closed or shrinking.
+		p.Unlock()
+		if resource != nil {
+			resource.Close()
+		}
+		return
+	}
+
+	if resource == nil {
+		p.Unlock()
+		p.ensureMinimumActive()
+		return
+	}
+
+	if p.state.InUse < 0 {
+		p.state.InUse++
+		p.Unlock()
+		panic(ErrPutBeforeGet)
+	}
+
+	w := resourceWrapper{resource: resource, timeUsed: time.Now()}
+	select {
+	case p.pool <- w:
+		p.state.InPool++
+	default:
+		// We don't have room in the pool.
+		p.state.InUse++
+		p.Unlock()
+		panic(ErrFull)
+	}
+
+	p.Unlock()
+}
+
+// SetCapacity changes the capacity of the pool.
+// You can use it to shrink or expand, but not beyond
+// the max capacity. If the change requires the pool
+// to be shrunk and `block` is true, SetCapacity waits
+// till the necessary number of resources are returned
+// to the pool.
+//
+// A SetCapacity of 0 is equivalent to closing the pool.
+func (p *FastPool) SetCapacity(capacity int, block bool) error {
+	p.Lock()
+
+	if p.state.Closed {
+		p.Unlock()
+		return ErrClosed
+	}
+
+	if capacity < 0 || capacity > cap(p.pool) {
+		p.Unlock()
+		return fmt.Errorf("capacity %d is out of range", capacity)
+	}
+
+	if capacity > 0 {
+		minActive := p.state.MinActive
+		if capacity < minActive {
+			p.Unlock()
+			return fmt.Errorf("minActive %v would now be higher than capacity %v", minActive, capacity)
+		}
+	}
+
+	if capacity == 0 {
+		p.state.Closed = true
+	}
+
+	isGrowing := capacity > p.state.Capacity
+	p.state.Capacity = capacity
+
+	if isGrowing {
+		p.state.Draining = false
+		p.Unlock()
+		return nil
+	}
+
+	p.state.Draining = true
+	p.Unlock()
```

Comment:
It looks like every return path prior to this point triggers an Unlock(). Tracking Unlock()s manually across non-trivial complexity code can be difficult, but you could pull this code into its own helper method and safely use a `defer p.Unlock()`.

Reply:
`shrink()` still needs to be unlocked, but I can take a look at `shrink()` so it can be called with the lock active.

Reply:
`shrink()` isn't called prior to this though? I'm suggesting pulling the whole previous block out into a method with a single `Lock() / defer Unlock()` rather than having explicit unlocks at every return point.

Reply:
Sorry, I was too terse in my response! I do understand what you mean, but I my plan to reduce the lock calls inside `shrink()` so I can just use a `defer Unlock()` in `SetCapacity` anyway. The  idea was to reverse the lock state requirement inside the `shrink()` call to be locked instead of unlocked at the entry point. It makes the `shrink()` code slightly easier to read with one less unlock call, and not within an `if` branch. I hope I'm making sense! 😄

I did consider to also do a helper method, but doesn't seem necessary with the change I made.

Diff: https://github.com/gak/vitess/commit/363b400a2e59e142f9097da54653388359ca8d85